nodes/move_to_pregrasp.py

```diff
@@ -0,0 +1,952 @@
+        goal_lift_near_base: bool,
+        init_length_near_mast: bool,
+    ) -> List[List[MoveToPregraspState]]:
+        """
+        Get the default state machine.
+
+        Parameters
+        ----------
+        horizontal_grasp: Whether the robot will be grasping the object horizontally
+            (True) or vertically (False).
+        init_lift_near_base: Whether the robot's arm is near the base at the start.
+        goal_lift_near_base: Whether the robot's arm should be near the base at the end.
+        init_length_near_mast: Whether the robot's arm length is near the mast at the start.
+
+        Returns
+        -------
+        List[List[MoveToPregraspState]]: The default state machine. Each list of states
+            (axis 0) will be executed sequentially. Within a list of states (axis 1), the
+            states will be executed in parallel.
+        """
+        states = []
+        # If the current arm lift is near the base, and the length is not already near the mast,
+        # move the arm to the stow height before fully stowing the arm length. This is to account
+        # for the case where the wrist is vertically down and may collide with the base.
+        if init_lift_near_base and not init_length_near_mast:
+            states.append([MoveToPregraspState.STOW_ARM_LENGTH_PARTIAL])
+            states.append([MoveToPregraspState.STOW_ARM_LIFT])
+            states.append([MoveToPregraspState.STOW_ARM_LENGTH_FULL])
+        else:
+            states.append([MoveToPregraspState.STOW_ARM_LENGTH_FULL])
+        states.append([MoveToPregraspState.STOW_WRIST])
+        # If the goal arm lift is near the base and we haven't already stowed the arm lift, stow the arm lift.
+        if goal_lift_near_base and MoveToPregraspState.STOW_ARM_LIFT not in states:
+            states.append([MoveToPregraspState.STOW_ARM_LIFT])
+        states.append([MoveToPregraspState.ROTATE_BASE, MoveToPregraspState.HEAD_PAN])
+        # If the goal is near the base and we're doing a vertical grasp, lengthen the arm before deploying the wrist.
+        if goal_lift_near_base and not horizontal_grasp:
+            states.append([MoveToPregraspState.LENGTHEN_ARM])
+            states.append([MoveToPregraspState.MOVE_WRIST])
+            states.append([MoveToPregraspState.LIFT_ARM])
+        else:
+            states.append([MoveToPregraspState.LIFT_ARM])
+            states.append([MoveToPregraspState.MOVE_WRIST])
+            states.append([MoveToPregraspState.LENGTHEN_ARM])
+        states.append([MoveToPregraspState.TERMINAL])
+
+        return states
+
+    def use_ik(self) -> bool:
+        """
+        Whether the state requires us to compute the arm's IK.
+
+        Returns
+        -------
+        bool: Whether the state requires the inverse jacobian controller.
+        """
+        return self in [
+            MoveToPregraspState.LIFT_ARM,
+            MoveToPregraspState.LENGTHEN_ARM,
+        ]
+
+
+class MoveToPregraspNode(Node):
```

Comment:
Do you see any performance hit from using so many ROS nodes?

Reply:
This feature adds one node (`/move_to_pregrasp`). The complete list of running nodes is:
```
/amcl
/behavior_server
/bt_navigator
/bt_navigator_navigate_through_poses_rclcpp_node
/bt_navigator_navigate_to_pose_rclcpp_node
/camera
/configure_video_streams
/controller_server
/d435i_accel_correction_node
/global_costmap/global_costmap
/gripper_camera
/joint_state_publisher
/joint_trajectory_action
/laser_filter
/laser_filter
/laser_filter
/lifecycle_manager_localization
/lifecycle_manager_navigation
/local_costmap/local_costmap
/map_server
/move_to_pregrasp
/navigation_camera
/planner_server
/robot_state_publisher
/rosapi
/rosapi_params
/rosbridge_websocket
/sllidar_node
/static_transform_publisher_3cyXiogXh3YjhR0C
/stretch_driver
/tf2_web_republisher
/transform_listener_impl_594129e43ba0
/transform_listener_impl_5d048bf95800
/transform_listener_impl_5fe79ce6f040
/transform_listener_impl_61f99bae9710
/transform_listener_impl_63b81553c220
/waypoint_follower
```

I haven't found any performance hit with the "low resolution" D435i. I did find a performance hit when I increased resolution of the D435i, presumably because both `/configure_video_streams` and `/move_to_pregrasp` subscribe to its streams. To address that, I did try to combine those two nodes into one, but that resulted in a priority inversion issue where action calls were not received for very long, presumably because camera messages were always ready to pop off of the waitlist before the action invokation.

Since we're switching to a python-backend web interface, I don't think it's worthwhile overfitting to ROS issues (as long as it doesn't hard functionality/usability/reliability).